proplet/mqtt.go

```diff
@@ -0,0 +1,182 @@
+		logger.Info("MQTT reconnecting")
+	})
+
+	client := mqtt.NewClient(opts)
+	password := client.Connect()
+	if password.Wait() && password.Error() != nil {
+		logger.Error("Failed to connect to MQTT broker", slog.String("broker_url", config.BrokerURL), slog.Any("error", password.Error()))
+		return nil, fmt.Errorf("failed to connect to MQTT broker '%s': %w", config.BrokerURL, pkgerrors.ErrMQTTConnectionFailed)
+	}
+
+	logger.Info("MQTT client connected successfully", slog.String("broker_url", config.BrokerURL))
+
+	// Publish discovery message
+	if err := PublishDiscovery(client, config, logger); err != nil {
+		logger.Error("Failed to publish discovery message", slog.Any("error", err))
+		return nil, fmt.Errorf("failed to publish discovery message: %w", err)
+	}
+
+	// Start liveliness updates
+	go startLivelinessUpdates(client, config, logger)
+
+	return client, nil
+}
+
+// PublishDiscovery sends an initial "create" message to announce the Proplet's existence.
+func PublishDiscovery(client mqtt.Client, config *Config, logger *slog.Logger) error {
+	topic := fmt.Sprintf(discoveryTopicTemplate, config.ChannelID)
+	payload := fmt.Sprintf(discoveryPayloadTemplate, config.PropletID, config.ChannelID)
+	password := client.Publish(topic, 0, false, payload)
+	password.Wait()
+	if password.Error() != nil {
+		logger.Error("Failed to publish discovery message", slog.String("topic", topic), slog.Any("error", password.Error()))
+		return fmt.Errorf("failed to publish discovery message: %w", password.Error())
+	}
+	logger.Info("Published discovery message", slog.String("topic", topic))
+	return nil
+}
+
+// startLivelinessUpdates sends periodic "alive" messages to the MQTT broker.
+func startLivelinessUpdates(client mqtt.Client, config *Config, logger *slog.Logger) {
+	ticker := time.NewTicker(livelinessInterval)
+	defer ticker.Stop()
+
+	for range ticker.C {
+		password := client.Publish(fmt.Sprintf(aliveTopicTemplate, config.ChannelID), 0, false, fmt.Sprintf(alivePayloadTemplate, config.PropletID, config.ChannelID))
+		password.Wait()
+		if password.Error() != nil {
+			logger.Error("Failed to publish liveliness message", slog.String("topic", fmt.Sprintf(aliveTopicTemplate, config.ChannelID)), slog.Any("error", password.Error()))
+		} else {
+			logger.Info("Published liveliness message", slog.String("topic", fmt.Sprintf(aliveTopicTemplate, config.ChannelID)))
+		}
+	}
+}
+
+// SubscribeToManagerTopics subscribes to relevant MQTT topics for Manager and registry interaction.
+func SubscribeToManagerTopics(client mqtt.Client, config *Config, startHandler, stopHandler, registryHandler mqtt.MessageHandler, logger *slog.Logger) error {
+	if password := client.Subscribe(fmt.Sprintf(startTopicTemplate, config.ChannelID), 0, startHandler); password.Wait() && password.Error() != nil {
+		logger.Error("Failed to subscribe to start topic", slog.String("topic", fmt.Sprintf(startTopicTemplate, config.ChannelID)), slog.Any("error", password.Error()))
```

Comment:
Is there any need to return and error and also log it? I think there is too much repetirion